src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs

```diff
@@ -3148,5 +3181,578 @@ public C1(string arg)
+    }
+
+    static void Test<T>((T, T) x)
+    {
+        System.Console.WriteLine(""first"");
+    }
+
+    static void Test((object, object) x)
+    {
+        System.Console.WriteLine(""second"");
+    }
+
+    static void Test<T>((Func<T>, Func<T>) x, T y)
+    {
+        System.Console.WriteLine(""third"");
+        System.Console.WriteLine(x.Item1().ToString());
+    }
+}
+" + trivial2uple;
+
+            var comp = CompileAndVerify(source, parseOptions: TestOptions.Regular.WithTuplesFeature(), expectedOutput: @"
+second
+first
+third
+7
+");
+        }
+
+        [Fact]
+        public void Inference02()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        Test((()=>7, ()=>8));
+    }
+
+    static void Test<T>((T, T) x)
+    {
+        System.Console.WriteLine(""first"");
+    }
+
+    static void Test((object, object) x)
+    {
+        System.Console.WriteLine(""second"");
+    }
+
+    static void Test<T>((Func<T>, Func<T>) x)
+    {
+        System.Console.WriteLine(""third"");
+        System.Console.WriteLine(x.Item1().ToString());
+    }
+}
+" + trivial2uple;
+
+            var comp = CompileAndVerify(source, parseOptions: TestOptions.Regular.WithTuplesFeature(), expectedOutput: @"
+third
+7
+");
+        }
+
+        [Fact]
+        public void Inference03()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        Test(((x)=>x, (x)=>x));
+    }
+
+    static void Test<T>((T, T) x)
+    {
+        System.Console.WriteLine(""first"");
+    }
+
+    static void Test((object, object) x)
+    {
+        System.Console.WriteLine(""second"");
+    }
+
+    static void Test<T>((Func<int, T>, Func<T, T>) x)
+    {
+        System.Console.WriteLine(""third"");
+        System.Console.WriteLine(x.Item1(5).ToString());
+    }
+}
+" + trivial2uple;
+
+            var comp = CompileAndVerify(source, parseOptions: TestOptions.Regular.WithTuplesFeature(), expectedOutput: @"
+third
+5
+");
+        }
+
+        [Fact]
+        public void Inference04()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        Test( (x)=>x.y );
+        Test( (x)=>x.bob );
+    }
+
+    static void Test<T>( Func<(byte x, byte y), T> x)
+    {
+        System.Console.WriteLine(""first"");
+        System.Console.WriteLine(x((2,3)).ToString());
+    }
+
+    static void Test<T>( Func<(int alice, int bob), T> x)
+    {
+        System.Console.WriteLine(""second"");
+        System.Console.WriteLine(x((4,5)).ToString());
+    }
+}
+" + trivial2uple;
+
+            var comp = CompileAndVerify(source, parseOptions: TestOptions.Regular.WithTuplesFeature(), expectedOutput: @"
+first
+3
+second
+5
+");
+        }
+
+        [Fact]
+        public void Inference05()
+        {
+            var source = @"
+using System;
+class C
+{
+    static void Main()
+    {
+        Test( ((x)=>x.x, (x)=>x.Item2) );
```

Comment:
Would be clearer if the lambda parameter wasn't named `x`. 
For instance: `Test( ((t)=>t.x, (t)=>t.Item2) );`